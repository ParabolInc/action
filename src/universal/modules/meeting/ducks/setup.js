import {Map as iMap, List as iList} from 'immutable';
import emailAddresses from 'email-addresses';
import * as _ from 'lodash';
import {fetchGraphQL} from '../../../utils/fetching';


// Setup 1
export const SETUP1_ADD_EMAILS_SUCCESS = 'action/meeting/SETUP1_ADD_EMAILS_SUCCESS';
export const SETUP1_ADD_EMAILS_ERROR = 'action/meeting/SETUP1_ADD_EMAILS_ERROR';
export const SETUP1_UPDATE_INVITES_FIELD = 'action/meeting/SETUP1_UPDATE_INVITES_FIELD';

// Setup 2
export const SETUP2_REMOVE_INVITEE = 'action/meeting/SETUP2_REMOVE_INVITEE';
export const SETUP2_UPDATE_ROW_HOVER = 'action/meeting/SETUP2_UPDATE_ROW_HOVER';
export const SETUP2_INVITES_ERROR = 'action/meeting/SETUP2_INVITES_ERROR';
export const SETUP2_INVITES_SUCCESS = 'action/meeting/SETUP2_INVITES_SUCCESS';

const initialState = iMap({
  setup1: iMap({
    emails: iList(),
    invitesField: '',
    invitesFieldHasValue: false,
    invitesFieldError: '',
    invitesFieldHasError: false
  }),
  setup2: iMap({
    rowWithHover: ''
  })
});

export default function reducer(state = initialState, action = {}) {
  switch (action.type) {
    case SETUP1_ADD_EMAILS_ERROR:
      return state.mergeDeep({
        setup1: iMap({
          invitesFieldError: 'invalid email addresses',
          invitesFieldHasError: true
        })
      });
    case SETUP1_ADD_EMAILS_SUCCESS:
      return state.mergeDeep({
        setup1: iMap({
          emails: iList(action.payload),
          invitesField: '',
          invitesFieldHasValue: false,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP1_UPDATE_INVITES_FIELD:
      return state.mergeDeep({
        setup1: iMap({
          invitesField: action.payload.value,
          invitesFieldHasValue: action.payload.hasValue,
          invitesFieldError: '',
          invitesFieldHasError: false
        })
      });
    case SETUP2_REMOVE_INVITEE: {
      const idx = action.payload;
      const list = state.getIn(['setup1', 'emails']);
      return state.mergeDeep({
        setup1: iMap({
          emails: [
            ...list.slice(0, idx),
            ...list.slice(idx + 1)
          ]
        })
      });
    }
    case SETUP2_UPDATE_ROW_HOVER:
      return state.mergeDeep({
        setup2: iMap({
          rowWithHover: action.payload
        })
      });
    default:
      return state;
  }
}

export const addInvitesFromInvitesField = (emailsString) => {
  let emails = emailAddresses.parseAddressList(emailsString);
  if (emails === null) {
    return ({
      type: SETUP1_ADD_EMAILS_ERROR,
      payload: null
    });
  }
  emails = _.map(emails, (em) => _.pick(em, ['name', 'address']));
  return ({
    type: SETUP1_ADD_EMAILS_SUCCESS,
    payload: emails
  });
};

export const updateInvitesField = (value) => {
  let hasValue = true;
  if (value === '') {
    hasValue = false;
  }
  return ({
    type: SETUP1_UPDATE_INVITES_FIELD,
    payload: {
      value,
      hasValue
    }
  });
};

export const removeInvitee = (nameOrEmail) =>
  (dispatch, getState) => {
    const idx = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails'])
      .findIndex((em) => em.name === nameOrEmail || em.address === nameOrEmail);
    return dispatch({
      type: SETUP2_REMOVE_INVITEE,
      payload: idx
    });
  };

// TODO: batch this query?
export const createInvites = (teamId) =>
  async (dispatch, getState) => {
    const queries = [];
    const invitees = getState()
      .getIn(['meetingModule', 'setup', 'setup1', 'emails']);
    for (const invitee of invitees) {
      const {address} = invitee;
      const name = invitee.name || '';
      const query = `
        mutation($teamId: ID!, $name: String!, $email: String!) {
          payload: createTeamMember(teamId: $teamId, name: $name, email: $email) {
            id,
            name,
            email
          }
        }`;
      queries.push(
        fetchGraphQL({query, variables: {teamId, name, email: address}})
      );
    }
    const results = await Promise.all(queries);
    const error = _.reduce(
      results,
      (agg, result) => (agg || result.error),
      false
    );
    if (error) {
      return dispatch({ type: SETUP2_INVITES_ERROR });
    }
    return dispatch({ type: SETUP2_INVITES_SUCCESS });
  };
