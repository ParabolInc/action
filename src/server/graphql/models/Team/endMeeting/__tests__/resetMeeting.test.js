import getRethink from 'server/database/rethinkDriver';
import MockDate from 'mockdate';
import {__now} from 'server/__tests__/setup/mockTimes';
import fetchAndTrim from 'server/__tests__/utils/fetchAndTrim';
import TrimSnapshot from 'server/__tests__/utils/TrimSnapshot';
import MockDB from 'server/__tests__/setup/MockDB';
import resetMeeting from 'server/graphql/models/Team/endMeeting/resetMeeting';

MockDate.set(__now);
console.error = jest.fn();

describe('resetMeeting', () => {
  test('resets the meeting state back to the lobby after the meeting ended', async() => {
    // SETUP
    const r = getRethink();
    const trimSnapshot = new TrimSnapshot();
    const mockDB = new MockDB();
    const {teamMember} = await mockDB.init()
      .newMeeting(undefined, {inProgress: true});
    const teamId = mockDB.context.team.id;
    const teamMemberIds = teamMember.filter((tm) => tm.teamId === teamId).map(({id}) => id);

    // TEST
    jest.useFakeTimers();
    const promise = resetMeeting(teamId);
    jest.runTimersToTime(5000);
    await promise;
    // VERIFY
    const db = await fetchAndTrim({
      agendaItem: r.table('AgendaItem').getAll(teamId, {index: 'teamId'}).orderBy('teamMemberId'),
      project: r.table('Project').getAll(r.args(teamMemberIds), {index: 'teamMemberId'}).orderBy('teamMemberId'),
      team: r.table('Team').get(teamId),
      teamMember: r.table('TeamMember').getAll(teamId, {index: 'teamId'}).orderBy('preferredName'),

    }, trimSnapshot);
    expect(db).toMatchSnapshot();
  });
});
