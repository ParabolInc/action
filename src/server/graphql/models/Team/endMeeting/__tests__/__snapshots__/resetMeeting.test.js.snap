// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`resetMeeting resets the meeting state back to the lobby after the meeting ended 1`] = `
Object {
  "agendaItem": Array [
    Object {
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Agenda Item[2]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.997Z,
      "createdBy": 3,
      "id": 0,
      "isActive": false,
      "isComplete": true,
      "sortOrder": 2,
      "teamId": 6,
      "teamMemberId": 7,
      "updatedAt": 2017-01-01T06:59:59.998Z,
    },
    Object {
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Agenda Item[1]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.998Z,
      "createdBy": 4,
      "id": 1,
      "isActive": false,
      "isComplete": true,
      "sortOrder": 1,
      "teamId": 6,
      "teamMemberId": 8,
      "updatedAt": 2017-01-01T06:59:59.999Z,
    },
    Object {
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Agenda Item[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.999Z,
      "createdBy": 5,
      "id": 2,
      "isActive": false,
      "isComplete": true,
      "sortOrder": 0,
      "teamId": 6,
      "teamMemberId": 9,
      "updatedAt": 2017-01-01T07:00:00.000Z,
    },
  ],
  "project": Array [
    Object {
      "agendaId": 2,
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Project[0]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.999Z,
      "createdBy": 5,
      "id": 10,
      "status": "active",
      "tags": Array [
        "private",
      ],
      "teamId": 6,
      "teamMemberId": 9,
      "updatedAt": 2017-01-01T07:00:00.000Z,
      "userId": 5,
    },
    Object {
      "agendaId": 1,
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Project[1]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.998Z,
      "createdBy": 4,
      "id": 11,
      "status": "active",
      "tags": Array [],
      "teamId": 6,
      "teamMemberId": 8,
      "updatedAt": 2017-01-01T06:59:59.999Z,
      "userId": 4,
    },
    Object {
      "agendaId": 0,
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Project[2]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.997Z,
      "createdBy": 3,
      "id": 12,
      "status": "active",
      "tags": Array [
        "private",
      ],
      "teamId": 6,
      "teamMemberId": 7,
      "updatedAt": 2017-01-01T06:59:59.998Z,
      "userId": 3,
    },
    Object {
      "agendaId": 0,
      "content": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"Test Project[3]\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
      "createdAt": 2017-01-01T06:59:59.996Z,
      "createdBy": 3,
      "id": 13,
      "status": "active",
      "tags": Array [],
      "teamId": 6,
      "teamMemberId": 7,
      "updatedAt": 2017-01-01T06:59:59.997Z,
      "userId": 3,
    },
  ],
  "team": Object {
    "activeFacilitator": null,
    "checkInGreeting": Object {
      "content": "Shalom",
      "language": "Hebrew",
    },
    "checkInQuestion": "{\\"entityMap\\":{},\\"blocks\\":[{\\"key\\":0,\\"text\\":\\"What’s something you’ve learned recently?\\",\\"type\\":\\"unstyled\\",\\"depth\\":0,\\"inlineStyleRanges\\":[],\\"entityRanges\\":[],\\"data\\":{}}]}",
    "facilitatorPhase": "lobby",
    "facilitatorPhaseItem": null,
    "id": 6,
    "isArchived": false,
    "isPaid": true,
    "meetingId": null,
    "meetingPhase": "lobby",
    "meetingPhaseItem": null,
    "name": "Team America",
    "orgId": 14,
    "tier": "personal",
  },
  "teamMember": Array [
    Object {
      "checkInOrder": 20,
      "email": "natasha.romanova@avengers.org",
      "id": 15,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Black Widow",
      "teamId": 6,
      "userId": 28,
    },
    Object {
      "checkInOrder": 21,
      "email": "steve.rogers@avengers.org",
      "id": 16,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": true,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Captain America",
      "teamId": 6,
      "userId": 29,
    },
    Object {
      "checkInOrder": 22,
      "email": "bruce.banner@avengers.org",
      "id": 17,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Hulk",
      "teamId": 6,
      "userId": 30,
    },
    Object {
      "checkInOrder": 23,
      "email": "tony.stark@avengers.org",
      "id": 18,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Iron Man",
      "teamId": 6,
      "userId": 31,
    },
    Object {
      "checkInOrder": 24,
      "email": "kyle.richmond@avengers.org",
      "id": 19,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Nighthawk",
      "teamId": 6,
      "userId": 32,
    },
    Object {
      "checkInOrder": 25,
      "email": "thor@avengers.org",
      "id": 9,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Thor",
      "teamId": 6,
      "userId": 5,
    },
    Object {
      "checkInOrder": 26,
      "email": "vision@avengers.org",
      "id": 8,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "Vision",
      "teamId": 6,
      "userId": 4,
    },
    Object {
      "checkInOrder": 27,
      "email": "james.rhodes@avengers.org",
      "id": 7,
      "isCheckedIn": null,
      "isFacilitator": false,
      "isLead": false,
      "isNotRemoved": true,
      "picture": "https://placeimg.com/100/100/animals",
      "preferredName": "War Machine",
      "teamId": 6,
      "userId": 3,
    },
  ],
}
`;
